File: src/main/java/be/superjoran/mint/domain/BankAccount.java
package be.superjoran.mint.domain;

import org.hibernate.annotations.Formula;

import javax.persistence.Access;
import javax.persistence.AccessType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.Table;
import java.math.BigDecimal;

/**
 * Created by Jorandeboever
 * Date: 01-Oct-16.
 */
@Entity
@Table(name = "T_BANKACCOUNT")
public class BankAccount extends DomainObject {
    private static final long serialVersionUID = -4411985824034001947L;

    @ManyToOne
    @JoinColumn(name = "OWNER_UUID")
    private Person owner;

    private Bank bank;

    @Column(nullable = false)
    private String number;

    @Column
    private String name;

    @Formula(value = "(SELECT SUM(statement.AMOUNT) FROM T_STATEMENT statement WHERE statement.ORIGINATINGACCOUNT_UUID = UUID)")
    @Access(AccessType.FIELD)
    private BigDecimal balance;

    public BankAccount() {
    }

    public BankAccount(Person owner) {
        this.owner = owner;
    }

    public Person getOwner() {
        return this.owner;
    }

    public void setOwner(Person owner) {
        this.owner = owner;
    }

    public Bank getBank() {
        return this.bank;
    }

    public void setBank(Bank bank) {
        this.bank = bank;
    }

    public String getNumber() {
        return this.number;
    }

    public void setNumber(String number) {
        this.number = number;
    }

    public String getName() {
        return this.name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public BigDecimal getBalance() {
        return this.balance;
    }

    public void setBalance(BigDecimal balance) {
        this.balance = balance;
    }

    @Override
    public String getDisplayValue() {
        return this.toString();
    }

    @Override
    public String toString() {
        return this.name != null ? this.name : this.number;
    }
}

File: src/main/java/be/superjoran/mint/domain/BankAccount.kt
package be.superjoran.mint.domain

import org.hibernate.annotations.Formula
import java.math.BigDecimal
import javax.persistence.*
import javax.validation.constraints.NotNull
import javax.validation.constraints.Pattern

/**
 * Created by Jorandeboever
 * Date: 01-Oct-16.
 */
@Entity
@Table(name = "T_BANKACCOUNT")
class BankAccount
(
        @ManyToOne
        @JoinColumn(name = "OWNER_UUID")
        var owner: Person,

        @Column(nullable = false)
        @get:NotNull @get:Pattern(regexp = "[A-Z]{2}[0-9]{2}(\\s[0-9]{4}){3}")
        var number: String
): DomainObject() {

    var bank: Bank = Bank.BELFIUS

    @Column
    var name: String? = null

    @Formula(value = "(SELECT SUM(statement.AMOUNT) FROM T_STATEMENT statement WHERE statement.ORIGINATINGACCOUNT_UUID = UUID)")
    @Access(AccessType.FIELD)
    var balance: BigDecimal? = null

    constructor(owner: Person, bank: Bank, number: String) : this(owner, number) {
        this.bank = bank
    }

    constructor(owner: Person, bank: Bank, number: String, name: String?) : this(owner, bank, number) {
        this.name = name
    }

    override val displayValue: String
        get() = if (this.name != null) this.name.toString() else this.number

    override fun toString(): String {
        return displayValue
    }
}
